CONTRIBUTING.md

```diff
@@ -71,6 +75,72 @@ IntelliJ's code generation can be configured to do this automatically:
      Tick:
      - Make generated local variables final
      - Make generated parameters final
+     
+**Null checking of method or function parameters**
+
+If arguments to a constructor set members of the object, or if arguments to a method
+change internal state, e.g. adding an item to a list, they should be checked for null, e.g.
+
+```
+public class MyObject {
+
+    private final String name;
+    private final List<String> myList = new ArrayList<>();
+
+    public MyObject(final String name) {
+        this.name = Objects.requireNonNull(name, "name");
+    }
+
+    public void addItem(final String item) {
+        myList.add(Objects.requireNotNull(item, "item"));
+    }
+}
+
+```
+
+**No TODOs in code**
+
+Please don't leave any TODOs in the code
+
+**New line for each method or constructor parameter**
+
+Methods or constructors should be defined with each parameter on a new line.
```

Comment:
I'd argue that just breaking long lines onto the next one, rather than one param per line, is sufficient, and is visually more appealing. I.e. let's just follow standard google style guide, I don't think this justifies an exception.

Reply:
🤔 I've grown to really like it because I can line up things side by side, count from the end or from the beginning easily (i..e the third last parameter). But I suppose it does just boil down to preference on this one, so maybe it makes sense to just follow google standards.

Reply:
Like you, I didn't like this when I first joined, but have got used to it.  

It does have the benefit of helping with our line width restriction and also it does often mean smaller diff when adding parameters to methods at the start or middle of the param list. The diff is only the new param, rather than the diff including all the params as they've all moved.

Personally, given most of the code is styled this way and most of us are used to it, I see no issue with this.  I'd of though it would be possible to configure the IDE / CheckStyle to support this, but haven't tried.

It may be worth spending some time to see if we can get tooling support for this style. If we do, we stick with it.

Reply:
I think if it's controversial we should err on the side of fewer rules (especially if it can't easily be enforced by tooling). So in this case, even though I like this rule, I think maybe we should ditch it.

Reply:
Another reason I don't like this. I want to be able to hit reformat in IntelliJ (or even better, auto reformat on save) and know that the resulting code will pass the project code format style so I don't have to worry about it, but I don't think there is any way to configure IntelliJ to format params, one on each line. So you have to do this manually which is a real pain.

Reply:
> If people like putting one param per line, I'd say that's fine. But I don't think we should mandate

If we go this route we risk code ping-ponging from one style to the next as each person that works in a file reformats it to their preferred style. This makes reviewing, merging and branching more difficult and is one of the main driving forces behind having cod styles in the first place.

Reply:
> If we go this route we risk code ping-ponging from one style to the next as each person that works in a file reformats it to their preferred style. This makes reviewing, merging and branching more difficult and is one of the main driving forces behind having cod styles in the first place.

I don't think this is the case as both one param per line and multiple params per line will pass company (Google) standard. Sometimes it might make sense to have one param per line (e.g. with annotations) for clarity. But in many cases it doesn't. The point here is not to mandate it.